src/charm.py

```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
+            userlist = self._read_file(USERLIST_PATH)
+            return pgb.parse_userlist(userlist)
         except FileNotFoundError:
-            # There is no existing userlist.txt file, so return an empty dict
-            return {}
+            logger.error("userlist not found")
 
-        parsed_userlist = {}
-        for line in userlist.split("\n"):
-            if line.strip() == "":
-                continue
-            # Userlist is formatted "{username}" "{password}""
-            username, password = line.replace('"', "").split(" ")
-            parsed_userlist[username] = password
+    def _reload_pgbouncer(self) -> None:
+        """Reloads pgbouncer application.
 
-        return parsed_userlist
+        Pgbouncer will not apply configuration changes without reloading, so this must be called
+        after each time config files are changed.
+        """
+        self.unit.status = MaintenanceStatus("Reloading Pgbouncer")
+        logger.info("reloading pgbouncer application")
 
-    def _generate_password(self) -> str:
-        """Generates a secure password.
+        pgb_container = self.unit.get_container(PGB)
+        if not pgb_container.can_connect():
+            self.unit.status = WaitingStatus(
+                "unable to restart pgbouncer - container not connected."
+            )
+            return
+        pgb_container.restart(PGB)
 
-        Returns:
-            A random 24-character string of letters and numbers.
-        """
-        choices = string.ascii_letters + string.digits
-        return "".join([secrets.choice(choices) for _ in range(24)])
+        self.unit.status = ActiveStatus("PgBouncer Reloaded")
 
-    def _hash(self, string: str) -> str:
-        """Returns a hash of the given string.
+    # =====================
+    #  K8s Charm Utilities
+    # =====================
 
-        Currently only implements md5
-        """
-        return hashlib.md5(string.encode()).hexdigest()
+    def _read_file(self, filepath: str) -> str:
+        """Reads file from pgbouncer container as a string.
 
-    # =================================================
-    #  PgBouncer service management / health functions
-    # =================================================
+        Args:
+            filepath: the filepath to be read
 
-    def _reload_pgbouncer(self) -> None:
-        """Reloads pgbouncer application.
+        Returns:
+            A string containing the file located at the given filepath.
 
-        Pgbouncer will not apply configuration changes without reloading, so this must be called
-        after each time config files are changed.
-        TODO implement function stub
+        Raises:
+            FileNotFoundError: if there is no file at the given path.
         """
-        logger.info("reloading pgbouncer application")
-        pass
+        pgb_container = self.unit.get_container(PGB)
+        inaccessible = f"pgbouncer container not accessible, cannot find {filepath}"
```

Comment:
As this variable is only relevant to the inner scope (`if`), it can be moved to inside there.